### ISSUE TITLE
Feature request: Replacing secrets into config

### ISSUE THREAD

**Issue by JacobSnyder** [linuxserver/docker-ddclient#44]
[linuxserverurl]: https://linuxserver.io
[![linuxserver.io](https://raw.githubusercontent.com/linuxserver/docker-templates/master/linuxserver.io/img/linuxserver_medium.png)][linuxserverurl]

------------------------------

The ddclient config, due to its nature, will inevitably have secrets in it (emails, usernames, passwords, tokens, etc.).  I like to check my docker-compose.yml and configs in to source control for backups and to share them, so I try to keep secrets out of them and instead leverage Docker's secrets whenever I can.  It would be nice if I could put tokens into my config file that would be replaced with secrets at runtime so I could check my config into source control and share it without concern.  

While it would be a "simple" solve to treat the entire ddclient.conf as a secret and not include it in source control, this leaves the entire config as unversioned and unshareable.

My proposal to an end-user is as follows:
- Expose Docker secrets to the container.  Ex: `cloudflare_email` and `cloudflare_api_key`.
- The linuxserver.io images already helpfully support populating environment variables with the contents of files via the `FILE__` prefix.  So, set `FILE__` environment variables to read these secrets into "real" environment variables.  Ex: `FILE__CLOUDFLARE_EMAIL=/run/secrets/cloudflare_email` and `FILE__CLOUDFLARE_API_KEY=run/secrets/cloudflare_api_key` would result in `CLOUDFLARE_EMAIL` and `CLOUDFLARE_API_KEY` being set to those secrets.
- Use a format understandable by [`envsubst`](https://www.gnu.org/software/gettext/manual/html_node/envsubst-Invocation.html) to author the config file.  Ex:
```
protocol=cloudflare,
zone=mydomain.com,
EMAIL=$CLOUDFLARE_EMAIL
password=$CLOUDFLARE_API_KEY
mydomain.com,mysubdomain.mydomain.com
```

In order to support this, the container would have to:
- Install `gettext` to get `envsubst`.
- Replace the config file copies in [the init script 30-config](https://github.com/linuxserver/docker-ddclient/blob/master/root/etc/cont-init.d/30-config) and [the service inotify_modify](https://github.com/linuxserver/docker-ddclient/blob/master/root/etc/services.d/inotify_modify/run) to `envsubst < /config/ddclient.conf > /ddclient.conf` instead of `cp`.
  - Potentially consider helping end-users debug this by exposing the modified config to them in a documented and visible manner.
  - Potentially consider helping end-users avoid issues by trimming trailing newlines from secrets.  Most text editors (nano and vim at least) will add a newline at the end of files.  This will result in `envsubst` putting unintentional newlines in the config file, which might cause ddclient's parsing some issues.

Please let me know if there's any interest in something like this, and please suggest improvements to my proposal - I'm pretty new to all of this stuff and understand that there's best practices that I'm not aware of.  If there's interest for this feature, and depending on the scope of the modifications to my proposal, I'd be happy to submit a change!  I already have a working proof-of-concept that I'm running via a private Dockerfile I've applied to this container to support this and my other feature request (#45).  If there isn't interest in having this as a feature, please advise if there would be interest in having this as a mod.

Thank you so much for your time and consideration and for maintaining such an awesome and comprehensive collection of images!

**Comment by derickson2402** [comment]
Hi, I have finally gotten back to fixing my copy of this repo that I use on my machines, and I found this issue when deciding whether to build my own container or use this one.

@JacobSnyder I agree with you, and I have an idea if you would find this useful as well. Usually in a containerized environment, secrets and configs are handled with environment variables which Docker passes to the container. These can be passed in file form as well:

```docker-compose.yml
version: "3.8"
services:
  ddclient:
    environment:
      - "ENV_VAR_1=foo"
      - "ENV_VAR_2=bar"
    env_file: "secrets.env"
```

with a corresponding file which reads:

```secrets.env
ENV_VAR_3=baz
ENV_VAR_4="Running out of env variable names..."
```

For my own projects where I have used this setup, I basically use ```entrypoint.sh``` in the Dockerfile to parse these settings at startup, which I believe is what most containerized programs do as well (although they probably do it better than in my projects). I would love to help on this project and submit a pull request if you would use this!

**Comment by tobbenb** [comment]
I'm not going to accept a PR to change the config int env variables only. First of all it would brake existing setups and second it would be way to many variables to cover all options in the config file and we have no time to check every release for new configuration options.
I think it's better to fork and make your own container if you want this behaviour.

I don't understand why a secrets file is better than using the config file? You can still secure it so only the user you choose has access to the file.

**Comment by derickson2402** [comment]
Sounds good, I will probably fork once I get some spare time, and I will try to remember to add a link here for anyone else looking.

In terms of system security, an env var file and a config file would be comparable in this case. I think the biggest thing is that env vars are a more secure and manageable way to introduce secrets in containerized programs, especially on something like k8's on AWS.

You have a good point though that existing configs would be broken, and since most use cases for DDNS are going to be on home systems and not on full scale k8's engines there is not any huge necessity to use env variables other than adhering to best practice. So definitely much safer to just fork.

**Comment by tobbenb** [comment]
It would be better to suggest supporting an .env file/reading env variables upstream. They should know all options in the config. Then we would not have any issue supporting it 😉

We don't test on swarm/k8s so if it works, great. If not, too bad.

**Comment by danepowell** [comment]
So what's the alternative? Is everyone really just shoving their creds into ddclient.conf? That's not tenable if you want to version control or share it.

Could we support the `-password` parameter for ddclient to set the password via an env var at runtime?

**Comment by JacobSnyder** [comment]
@danepowell 

As I mentioned in the original post, I'm running some customizations to this Docker image that address this, !41, and !45.  With the release of ddclient 3.10.0, the modifications from !45 aren't needed anymore.  Here's how I have things setup to address this and !41.

This leverages secrets, as documented here: https://docs.linuxserver.io/images/docker-hedgedoc#environment-variables-from-files-docker-secrets

# `docker-compose.yml`

Docker Compose does two main things.  First, it declares the secret for the API token and adds it to so that it will be exposed as the environment variable `CLOUDFLARE_API_TOKEN`.  Second, it defines how to build my customizations.

```
version: "3.8"

services:

  ddclient:
    # https://hub.docker.com/r/linuxserver/ddclient
    build:
      context: ./ddclient/build
      args:
        DDCLIENT_VERSION: v3.10.0-ls105 # 3f7a3ddf89fb9bd52f6189006cf2556365abfdbee4122573b2e85697bf10f924
    container_name: ddclient
    environment:
      PUID: 1000
      PGID: 1000
      FILE__CLOUDFLARE_API_TOKEN: /run/secrets/cloudflare_ddclient_api_token
    secrets:
      - cloudflare_ddclient_api_token
    volumes:
      - ./ddclient/config:/config
    restart: unless-stopped

secrets:
  cloudflare_ddclient_api_token:
    file: ./secrets/cloudflare-ddclient-api-token.secret
```

# `ddclient/build/Dockerfile`

The Dockerfile applies the patches necessary for my customizations.

```
ARG DDCLIENT_VERSION=latest
ARG GIT_BRANCH=master

FROM linuxserver/ddclient:${DDCLIENT_VERSION}

COPY root/ /

RUN \
    echo "**** install build packages ****" && \
    apk add --no-cache --virtual=build-dependencies \
        patch\
        && \
    echo "**** install runtime packages ****" && \
    apk add --no-cache \
        gettext \
        && \
    echo "**** patch base image ****" && \
    patch /etc/cont-init.d/30-config /tmp/30-config.patch && \
    patch /etc/services.d/inotify_modify/run /tmp/inotify_modify.patch && \
    echo "**** cleanup ****" && \
    apk del --purge  \
        build-dependencies \
        && \
    rm -rf \
        /tmp/*
```

# `ddclient/build/root/tmp/30-config.patch`

This patches the copying of the `ddclient.conf` file from the config directory to make it also substitute secrets with environment variables.

```
--- 30-config
+++ 30-config-patched
@@ -10,7 +10,9 @@
     cp /defaults/ddclient.conf /config

 # copy config from /config to root
-cp /config/ddclient.conf /ddclient.conf
+# cp /config/ddclient.conf /ddclient.conf
+# replace env vars into config from /config and write to root
+envsubst < /config/ddclient.conf > /ddclient.conf

 # permissions
 chown -R abc:abc \

```

# ``ddclient/build/root/tmp/inotify_modify.patch`

This patch fixes #41 and makes config file updates trigger secret substitution. 

```
--- run	2021-01-24 23:10:17.163084279 +0000
+++ run-patched	2021-01-24 23:08:17.000000000 +0000
@@ -1,10 +1,16 @@
 #!/usr/bin/with-contenv bash
 
 # starting inotify to watch /config/ddclient.conf and restart ddclient if changed.
-while inotifywait -e modify /config/ddclient.conf; do
-	cp /config/ddclient.conf /ddclient.conf
-	chown abc:abc /ddclient.conf
-	chmod 600 /ddclient.conf
-	s6-svc -h /var/run/s6/services/ddclient
-	echo "ddclient has been restarted"
-done
+inotifywait -mq -e close_write,moved_to --format "%e %f" /config/ |
+	while read events filename; do
+		if [ "$filename" != "ddclient.conf" ]; then
+			continue
+		fi
+
+		# cp /config/ddclient.conf /ddclient.conf
+		envsubst < /config/ddclient.conf > /ddclient.conf
+		chown abc:abc /ddclient.conf
+		chmod 600 /ddclient.conf
+		s6-svc -h /var/run/s6/services/ddclient
+		echo "ddclient has been restarted"
+	done
```

# `ddclient/config/ddclient.conf`

This looks pretty standard, except now it supports secret substitution using `envsubst` by looking for variables with the `$` prefix.  For example, my CloudFlare configuration now looks like:

```
protocol=cloudflare,
zone=mydomain.com,
password=$CLOUDFLARE_API_TOKEN
mydomain.com
```

# Running

Then, to make this whole thing work:

```
$ docker-compose build 
$ docker-compose up -d
```
